### PR TITLE
Fixed #24098 -- Added no-op attributes to RunPython and RunSQL

### PR DESCRIPTION
**Pull request by MarkusH**
https://code.djangoproject.com/ticket/24098

### CODE REVIEW

File: django/db/migrations/operations/special.py [hunk 1]
```diff
@@ -175,3 +178,5 @@ def database_backwards(self, app_label, schema_editor, from_state, to_state):
 
     def describe(self):
         return "Raw Python operation"
+
+    noop = lambda apps, schema_editor: None
```
Comment (loic) [inline django/db/migrations/operations/special.py]: Won't that create an unbound method on PY2? IMO we should just use `def` with `@staticmethod`.

File: docs/ref/migration-operations.txt [hunk 2]
```diff
@@ -245,6 +245,12 @@ operation that adds that field and so will try to run it again).
     The ability to pass parameters to the ``sql`` and ``reverse_sql`` queries
     was added.
 
+.. versionadded:: 1.8
+
+When you don't give a ``reverse_sql``, the operation and hence the migration
```
Comment (loic) [inline docs/ref/migration-operations.txt]: I'd rephrase to make it more obvious that what's `versionadded:: 1.8` is the `RunSQL.noop` part.

File: docs/ref/migration-operations.txt [hunk 3]
```diff
@@ -245,6 +245,12 @@ operation that adds that field and so will try to run it again).
...
+it belongs to, is not reversible. If you e.g. want to propagate a newly added
+column, you can set ``reverse_sql`` to ``RunSQL.noop``.
```
Comment (loic) [inline docs/ref/migration-operations.txt]: I'm not sure I understand the example.

File: docs/releases/1.8.txt [hunk 4]
```diff
@@ -457,6 +457,11 @@ Migrations
 * A :ref:`generic mechanism to handle the deprecation of model fields
   <migrations-removing-model-fields>` was added.
 
+* The :class:`~django.db.migrations.operations.RunPython` and
+  :class:`~django.db.migrations.operations.RunSQL` operations now have a
+  ``noop`` class attribute that can be used to mark respective forward or
+  backwards operations as no-ops.
```
Comment (loic) [inline docs/releases/1.8.txt]: I'd give an example here: e.g. "This is particularly useful to mark ``RunPython`` or ``RunSQL`` operations as reversible".

File: django/db/migrations/operations/special.py (hunk 1 above)
Comment (MarkusH) [inline django/db/migrations/operations/special.py]: Yes. Changed it.

**Comment by MarkusH** [comment]
Rephrased your doc suggestions.

File: django/db/migrations/operations/special.py [hunk 5]
```diff
@@ -100,9 +100,9 @@ def database_backwards(self, app_label, schema_editor, from_state, to_state):
     def describe(self):
         return "Raw SQL operation"
 
-    def _run_sql(self, schema_editor, sql):
-        if isinstance(sql, (list, tuple)):
-            for sql in sql:
+    def _run_sql(self, schema_editor, sqls):
```
Comment (loic) [inline django/db/migrations/operations/special.py]: I think I'd favor `noop` directly at this level rather than supporting `noop` within the list. So `sql=('', ('something', 'something')` triggers an error rather than just skipping as no-op.

File: django/db/migrations/operations/special.py [hunk 6]
```diff
@@ -110,12 +110,15 @@ def _run_sql(self, schema_editor, sql):
                         sql, params = sql
                     else:
                         raise ValueError("Expected a 2-tuple but got %d" % elements)
-                schema_editor.execute(sql, params=params)
-        else:
-            statements = schema_editor.connection.ops.prepare_sql_script(sql)
+                if sql != RunSQL.noop:
```
Comment (loic) [inline django/db/migrations/operations/special.py]: Wouldn't `is not` be more appropriate?

Comment (loic) [inline django/db/migrations/operations/special.py]: Actually maybe not, I had `noop == function` in mind but that's for `RunPython`.

File: tests/migrations/test_operations.py [hunk 7]
```diff
@@ -1620,6 +1643,18 @@ def inner_method(models, schema_editor):
...
+    def test_run_python_noop(self):
+        """
+        #24098 - Tests no-op RunPython operations.
+        """
+        project_state = ProjectState()
+        new_state = project_state.clone()
+        operation = migrations.RunPython(migrations.RunPython.noop, migrations.RunPython.noop)
+        with connection.schema_editor() as editor:
+            operation.database_forwards("test_runpython", editor, project_state, new_state)
+        with connection.schema_editor() as editor:
```
Comment (loic) [inline tests/migrations/test_operations.py]: Why is a new `with` statement needed?

**Comment by loic** [comment]
Left a couple more comments but patch is looking quite good already. I'll defer to @timgraham for the docs review.

**Comment by timgraham** [comment]
Edits: http://dpaste.com/0MTNJTC
The idea is to make the `versionadded` annotations self-contained so when we remove them, we either remove just that line or the entire block if its contents are indented. In you flow something like "The ``RunPython.noop`` static method was added" in a paragraph we are otherwise going to keep, then we have to reflow the text and it's more messy.

**Comment by MarkusH** [comment]
Thanks, @timgraham. Applied your changes.

**Comment by timgraham** [comment]
Merged in c8bac4b556cf4716dc9003e5da48060cb72ba7cb.
